test-data/unit/check-expressions.test

```diff
@@ -808,10 +808,39 @@ class B:
 class C:
     pass
 [out]
-main:3: error: Invalid index type "C" for "A"
+main:3: error: Invalid index type "C" for "A"; expected type "B"
 main:4: error: Incompatible types in assignment (expression has type "A", target has type "C")
 main:5: error: Unsupported target for indexed assignment
 
+[case testOverloadedIndexing]
+
+from typing import overload
+
+a, b, c = None, None, None  # type: (A, B, C)
+a[b]
+a[c]
+a[1]  # Fail
+
+i, s = None, None  # type: (int, str)
```

Comment:
It's okay to have some apparent redundancy, since indexing might hit some different code paths eventually. Having too many tests is less of a problem than not having enough :-)